### ISSUE TITLE
Syntax for simple port NAT

### ISSUE THREAD

**Issue by dkebler** [firehol/firehol#375]
Running latest firehol.

Just trying to do a simple NAT of ssh from port 322 on wan interface to an internal ssh server on port 22 ip 10.0.0.3

here is my conf file.  When I try it it hangs What am I doing wrong.  
FYI ssh to ssh server on gateway works fine, as well as dnat to internal https server only translating port doesn't work
```
# have outgoing traffic use the public IP
  ipv4 masquerade "${world}"

# nat http and https to internal server
dnat4 to 10.0.0.3   inface ${world} proto tcp  dport 80,443    # http,https to NAS Caddy Server
dnat4 to 10.0.0.3:22 inface ${world} proto tcp  dport 322   # ssh from outside to NAS Server
router4 internet2lan inface ${world} outface ${home} dst 10.0.0.0/24
        client all             accept
        server http,https      accept dst 10.0.0.3
        server tcp/322         accept dst 10.0.0.3

# LAN network facing interface
interface "${home}" home
   policy accept


# internet facing wan interface
interface "${world}" world
   protection bad-packets
   server ssh accept
   server tcp/322 accept
   server icmp accept
   client all accept
```

**Comment by philwhineray** [comment]
The dnat takes effect before the router rule, so replacing the `tcp/322` with ssh should work.

The `tcp/322` in interface world just opens that port on the server itself, so you probably want to remove that.

If, before doing the above changes, you look in the generated logs you should be able to see the ssh to 10.0.0.3 being blocked (it will list the destination port as 22).

The opposite is true for those that use snat. From the [nat](https://firehol.org/firehol-manual/firehol-nat/) manpage:


> When using dnat or redirect, the transformation is in the PREROUTING chain of the NAT table and happens before normal rules are matched, so your client or server rule should match the "modified" traffic.
>
> When using snat, the transformation is in the POSTROUTING chain of the NAT table and happens after normal rules are matched, so your client or server rule should match the "unmodified" traffic.

**Comment by dkebler** [comment]
ok what you suggested works as in port 22 goes to gateway box and port 322 goes to internal box.

Still I just don't grok why the router4 command doesn't end up sending traffic coming in on port 22 to internal box as well.  Does the router4 only affect traffic that is already made it to my "home" interface???  And the wan port 22 traffic never gets there?  Is that why this works correctly?

**Comment by philwhineray** [comment]
The first thing that happens to the packet when the server processes it, is that it changes the destination address and port.

Only then is a decision made about whether the packet is for a local `interface` or some other host via a `router`.

This link might help explain (or might make your eyes bleed, sorry if it's the latter):

https://upload.wikimedia.org/wikipedia/commons/3/37/Netfilter-packet-flow.svg

The key is to note that either traffic is sent to the local machine or it is sent back out again, never being handled by a process attached to the `wan` interface.

Your comments make me think you may also want to drop the `ssh` line from your wan interface, too since that permits ssh connections to the firewall from the outside, on regular port 22.

**Comment by philwhineray** [comment]
Sorry I could have been clearer. The 322 port packets get re-written so have the new IP, the regular port 22 do not get rewritten, so still have the IP of the gateway.

I also see you're aware that the ssh is exposed - sorry for the noise.

**Comment by dkebler** [comment]
Yes I want ssh to my gateway on port 22 which is running firehol.   

So is just opening up the port in the wan interface the best way or is it better(more secure/can I) NAT that to 10.0.0.1 (the gateway lan interface) and then route it back to 10.0.0.1??

I tried it and it didn't work.
```
# nat http and https to internal server
dnat4 to 10.0.0.3   inface ${world} proto tcp  dport 80,443    # http,https to NAS Caddy Server
dnat4 to 10.0.0.3:22 inface ${world} proto tcp  dport 322   # ssh from outside to NAS Server
dnat4 to 10.0.0.1 inface ${world} proto tcp  dport 22   # ssh from outside to NAS Server
router4 internet2lan inface ${world} outface ${home} dst 10.0.0.0/24
        client all             accept
        server http,https      accept dst 10.0.0.3
        server ssh             accept dst 10.0.0.1,10.0.0.3

# LAN network facing interface
interface "${home}" home
   policy accept


# internet facing wan interface
interface "${world}" world
   protection bad-packets
#  server ssh accept
   server icmp accept
   client all accept
```
# nat http and https to internal server
dnat4 to 10.0.0.3   inface ${world} proto tcp  dport 80,443    # http,https to NAS Caddy Server
dnat4 to 10.0.0.3:22 inface ${world} proto tcp  dport 322   # ssh from outside to NAS Server
dnat4 to 10.0.0.1 inface ${world} proto tcp  dport 22   # ssh from outside to NAS Server
router4 internet2lan inface ${world} outface ${home} dst 10.0.0.0/24
        client all             accept
        server http,https      accept dst 10.0.0.3
        server ssh             accept dst 10.0.0.1,10.0.0.3

# LAN network facing interface
interface "${home}" home
   policy accept


# internet facing wan interface
interface "${world}" world
   protection bad-packets
#  server ssh accept
   server icmp accept
   client all accept
```

**Comment by philwhineray** [comment]
Do you need to connect to it from the outside? If so open it on the wan as you did but bear in mind that people will attack it for sure. Adding a layer of NAT would not help even if it could be made to work.

Alternatively ssh onto your internal server from outside and then back to the gateway from the inside, although obviously there are more ways for this to go wrong if you amend the running firewall whilst connected in this manner.

**Comment by dkebler** [comment]
Yes I understand the risk.  I am using a key with passphrase.  My intention is to also filter on trusted ips.  I haven't implemented that.  I also was going to turn off icmp to make it harder to be seen.  I'm just doing preliminary testing at this point.

At the very least I'd like to filter on geolite2 country list like I do with Caddy.     That way I'd at least limit any attack from US only.  It would be better to do this at gateway firewall than at the webserver.  How do I refer to a file of ips as either whitelist or blacklist?  Is that explained in the docs somewhere, didn't see it.

BTW is it possible to filter on mac address of the inbound packet?  That would be the best.  Would allow sysop with known machine to be on any ip.

**Comment by dkebler** [comment]
looks like ipset is the way to enable large sets of ips for blacklist?
https://firehol.org/firehol-manual/firehol-ipset/
` ipv4 ipset addfile badguys file-with-the-bad-guys-ips.txt`
only if I do
```
/etc/firehol$ sudo ipset list -n
ipset v6.34: Kernel error received: Invalid argument
```
I thought I a got all ipfilter modules built in kernel.  Did I miss one that is needed for ipset

**Comment by philwhineray** [comment]
Yes, ipset is the most sensible way to handle many IPs or networks. It's also the best way to handle anything dynamic, since you do not have to reload the firewall to change the entries.

**Comment by fariouche** [comment]
Hello,
I know, it is a very old ticket, but as it is still open, I believe it is still not solved. In my case, I solved it using the dnat4 line above + the following:
```
router4 wan2lan eth1 world outface br0  dts "${lan_ips}"
    route custom forward tcp/322 default accept
```

in br0 I have eth0 (lan) and wifi0, with rules to block some traffic from wlan0 to eth0

I didn't had to open the tcp/322 port in my eth1 and br0 interface4 blocks!

Hope this helps, hopefully this is secure enough (except the ip filtering that need to be added too)
This may help having a working forward example in the doc maybe.

Thanks
